feat: Solid community server Docker image

matthieubosquet opened the pull request:
Follow up to the previous [PR #561 and comments](https://github.com/solid/community-server/pull/561). For some reason the branch has gone AWOL, hence the new PR, sorry about this.

It addresses the concerns raised by @ianconsolata about the use of `git clone` https://github.com/solid/community-server/pull/561#pullrequestreview-579690142

It also splits the image in two stages: one for the build, and one for the runtime; to make full use of Docker caching and ship a clean image containing only runtime files.

I also added an `npm audit` stage as a minimum check for potential vulnerabilities when building the image.

---

I gave a bit more thought to publishing the image to GitHub's container registry as an interim solution and I think we should hold of for now and wait until we manage to retrieve the DockerHub credentials because:
- It seems to me like there is no urgency since nothing stops people from building their own Docker images using this file as a base
- Having two places where the image can be found will probably create confusion rather than solve any problem
- The way Personal Access Tokens are scoped is suboptimal and they are required to deploy to GHCR

I would suggest we can reconsider in two weeks on the 15th if there has been no progress on the retrieving DockerHub credentials front.

Let me know what you think @ianconsolata, @pheyvaer & @RubenVerborgh. And thank you for your comments! ;)

ianconsolata commented on Dockerfile:
```diff
@@ -0,0 +1,49 @@
+FROM node:lts AS build
+
+## Create the solid community server directory
+RUN mkdir /community-server
```
The WORKDIR on the next line implicitly runs mkdir. :D

matthieubosquet replied:
OMG! Great find, thanks for the tip :)

ianconsolata commented on Dockerfile:
```diff
@@ -0,0 +1,49 @@
+## Copy runtime files from build stage
+COPY --from=build /community-server/package.json .
+COPY --from=build /community-server/bin ./bin
+COPY --from=build /community-server/config ./config
```
I think the config path is wrong. You create a dir at `/config` but then use `./config`. Since the WORKDIR is `/community-server` that path expands to `/community-server/config`. I assume you meant `/config`?

I also noticed you are only setting up the config and data directories for mounting. What about the templates directory? Is that something that has to be configured at build-time? If not, we may want to make a `/templates` as well to document it's mountable like data and config.

matthieubosquet replied:
I read somewhere that it was problematic to mount non empty directories from a windows host.
Hence the default two empty mount points (instead of a config mounting point with preloaded default configs).

I don't think the templates dir is really used for anything configurable at the moment. Even the `podTemplateFolder` CLI argument points to `templates/pod`. So I would rather not put it in for the sake of simplicity. I imagine this template folder and its use might evolve in the relatively near future. So I'd rather keep creating a dedicated mounting point for when the use case comes up, another nice little PR will make it possible, so I'm not too worried.

ianconsolata replied:
Oh interesting, I haven't heard that. Maybe call it `preset-configs` or something to make the distinction clear? In general, I like creating non-empty mount points because then it copies those files into the mounted dir, which is useful in Kubernetes and other orchestration platforms because it means the mounted volume starts off container the default config, including the default values. How will the Solid server handle having config spread across two directories? Are you expecting folks to copy the config files they need to the mounted directory from the preloaded folder, or just reference the preloaded files from the mounted folder? It should be fine either way, I'm mostly asking because my next step after this is merged is to use this container to deploy to Kubernetes and I want to make sure I line up with y'alls plans on how to configure things.

Makes sense regarding the template.

matthieubosquet replied:
If you are only going to use the included configurations, then you don't really need to mount a volume, you can just pass in arguments.

If using a custom config, you could indeed copy the config directory over and do as you please.

But you'll still be constrained in the sense that you could only use the base components that ship with the community server. For more complex use cases really taking advantage of the server's modularity, you'll need to install extra packages, run the build step and so on, so that wouldn't be covered by this image.

Which I think is fine, since the goal here is relative simplicity to cover all the most regular use cases and provide an example to get people started.

Then, for more advance uses, I'm sure people will have no issue either basing from this image and extending it or creating their own from scratch! ;)

ianconsolata commented on Dockerfile:
```diff
@@ -0,0 +1,49 @@
+ENTRYPOINT [ "node", "bin/server.js" ]
+
+## By default run in filemode (overriden if passing alternative arguments)
+CMD [ "-c", "config/config-file.json", "-f", "/data" ]
```
Conventionally, ENTRYPOINT and CMD come at the end. I don't know if there is a specific technical reason for it, but I find it much more readable.

matthieubosquet replied:
I thought it might influence caching, it doesn't ^^,
Put it back at the end.

ianconsolata commented on Dockerfile:
```diff
@@ -0,0 +1,49 @@
+# Runtime stage
+FROM node:lts
+
+## Add contact informations for questions about the container
```
There are a lot of comments like this that are self-evident for anyone accustomed to reading docker files, and a short google search away for the rest. I personally don't like including comments like that, because I think they just clutter the code.

If you do want to include them still to improve the experience for unfamiliar developers, I would recommended putting more spaces or some other indicator between the build and runtime stages. It's a little hard to tell this is a multi-stage build from first glance, because every line has spacing between it.

matthieubosquet replied:
I added two lines.

ianconsolata commented on Dockerfile:
```diff
@@ -0,0 +1,49 @@
+COPY . .
+
+## Verify if there are known vulnerabilities in the dependencies
+RUN npm audit --production --audit-level=high
```
Will this error and stop the build if it finds known vulnerabilities? Do you want it to?

matthieubosquet replied:
Yes and yes.

ianconsolata commented on Dockerfile:
```diff
@@ -0,0 +1,49 @@
+WORKDIR /community-server
+
+## Copy the dockerfile's context's community server files
+COPY . .
```
I think you still want to include the two stage package.json copy trick I mentioned in the other PR, as it will improve the build time of this image significantly. Otherwise the dependencies will be rebuilt every time there is a code change, even if there is no change in the dependencies. I got the suggestion from the Nodejs guide on Dockerizing node apps: https://nodejs.org/en/docs/guides/nodejs-docker-webapp/

matthieubosquet replied:
Good point!

ianconsolata reviewed (approved):
LGTM. There is one issue with the /config dir I think you probably want to address, but looks great otherwise. The two stage build was a great idea.

ianconsolata commented:
I tried using my own version of this built image, and it can't find the `dist` folder. How is that generated? Shouldn't `npm ci` handle that?

ianconsolata commented:
Ok, did some more digging and `npm ci` generates the `dist` folder on my local non-docker environment, but the CI build of the docker image shows the following warning:
```npm WARN lifecycle @solid/community-server@0.7.0~prepare: cannot run in wd @solid/community-server@0.7.0 npm run build (wd=/community-server)```

I'm not totally sure what's different about my CI setup, but adding `--unsafe-perm` to the `npm ci` line fixes things. I don't know enough about npm / node to know if that's the right solution or not, but wanted to pass along the issues I encountered.

matthieubosquet commented:
@ianconsolata really good spot!

Do you remember this line?
https://github.com/solid/community-server/blob/62610bbf6fd52f9e78d6f77be9be0d973f8c1827/Dockerfile#L11

I was confused as to why it needed to be added to the Dockerfile.

The answer can be found by running the Docker build with full output:
```bash
$ docker build --progress=plain --rm -f ./Dockerfile -t solid-community-server:latest .
...
#13 [build 6/6] RUN npm ci
...
#13 22.23 husky > Setting up git hooks
#13 22.25 Husky requires Git >=2.13.0. Got v2.11.0.
#13 22.25 husky > Failed to install
...
```

The node lts image comes with a lower version of Git than what is required by Husky to setup the post-install hooks. So, the build doesn't run ^^,

My build started working without it because of `COPY . .`... and running it on a local prebuilt CSS.

I added `npm run build` back in and went for the `node:lts-alpine` image by default since it makes the image 275MB vs 1.1GB when using `node:lts`.

ianconsolata commented:
Oh, interesting. Good call on switching to alpine, but I'm a little uncertain what the fix was. Is it calling `npm run build` instead of `npm ci` or does it just need to use an upgraded version of git with `npm ci`?

matthieubosquet commented:
@ianconsolata mmmh, you're right, it's not husky failing, the post-install build is run via the `prepare` script.

Well, I didn't find a very clear explanation as to why `npm prepare` supposedly runs into some form of user permissions issues. So, I'm if anything as confused as you.

I'd much rather run the build step explicitly than use the `--unsafe-perm` flag though. But I don't know enough about npm to give a clear answer here. I might try to get to the bottom of it later.

RubenVerborgh commented:
@matthieubosquet Is this a problem of root versus non-root? https://docs.npmjs.com/cli/v6/using-npm/config#unsafe-perm

RubenVerborgh commented on Dockerfile:
```diff
@@ -0,0 +1,49 @@
+COPY . .
+
+## Install and build the Solid community server (prepare script cannot run in wd)
+RUN npm ci && npm run build
```
```suggestion
RUN npm ci --ignore-scripts && npm run build
```

matthieubosquet commented:
@RubenVerborgh you're probably right. I'm just a bit perplex at what is happening in therms of UID switching with the "triggered" vs explicitly run scripts.

matthieubosquet commented:
@joachimvh, @RubenVerborgh, @ianconsolata I just reworked a bit the Readme part of the Docker and removed the CI actions related to publishing the Docker image to GitHub container registry since we decided not to use that.

I would say, let's merge that since I think the Dockerfile is perfectly fine and let's add CI for publishing images to DockerHub when we get hold of the account.

ianconsolata commented:
@matthieubosquet I think Ruben is probably right about the permissions issue being the cause, given that my addition of unsafe-perm fixed it. Here is the thread I used when debugging the issue, and I think it looks like there might be a couple solutions besides using unsafe-perm: https://github.com/npm/npm/issues/3497 Specifically, I think the issue is caused by NPM trying to downgrade it's privileges to non-root, and then trying to run in `/community-server` where it doesn't have privileges. It may be solved if we used a non-root user in the docker image, and `chmod` the `/community-server` directory so it worked with that user. It may also work if we pick a directory that is intended for user permissions already. I am still confused why the local build and the CI build run docker with different users, but I'm assuming that just has to do with CI and your local machine running docker with different permissions.

Merging the docker container without the CI sounds great to me! Having the docker image on master is useful for new contributors and for documentation even if it's not auto-publishing to dockerhub yet.

matthieubosquet commented:
@ianconsolata thanks for the link! Also, just to double check, the fix implemented here is line 17 of the Dockerfile: `RUN npm ci --ignore-scripts && npm run build`, does it work in your environment? I double checked in mine where all is fine.

@RubenVerborgh or @joachimvh do you have permission to cancel and rerun the pipeline, it seems to be hanging: https://github.com/solid/community-server/actions/runs/610114819

And let's merge once it passes.

joachimvh commented:
> @RubenVerborgh or @joachimvh do you have permission to cancel and rerun the pipeline, it seems to be hanging: https://github.com/solid/community-server/actions/runs/610114819

I should, but for some reason I get `Failed to cancel workflow. ` here, so something seems to be broken. Perhaps you can try force pushing again to restart that way?

matthieubosquet commented:
@joachimvh and that worked :)

ianconsolata commented:
Change worked great! Thanks for merging.
